File: Rotated_Sorted_Min.py
class Solution(object):
    def findMin(self, nums):
        """
        Solution:
        1. One chunk will definitely be unsorted, go to that chunk
        2. Do this until you have exactly 1 or 2 elements remaining
        3. Return the minimum of those remaining elements

        -- Passed all test cases on Leetcode.
        """

        # O(logn) Time Complexity | O(1) Space Complexity
        low = 0
        high = len(nums) - 1

        while (low <= high):
            mid = low + int((high - low) / 2)
            if (nums[mid] < nums[high]):
                if (low == mid):
                    return nums[mid]
                high = mid
            elif (nums[mid] > nums[low]):
                if (high == mid):
                    return nums[mid]
                low = mid
            else:
                return min(nums[low], nums[high])

File: src/com.madhurima/RotatedSortedMin.java
/*
    Time Complexity = O(logN)
    Space Complexity = O(1)
    Did this code successfully run on Leetcode : yes
 */

package com.madhurima;

public class RotatedSortedMin {
}

class Solution3 {
    public int findMin(int[] nums) {
        if(nums.length == 1){
            return nums[0];
        }

        int n = nums.length;
        int low = 0;
        int high = n-1;

        while(low <= high){
            if(nums[low] <= nums[high]){
                return nums[low];
            }

            int mid = low + (high - low)/2; //to prevent integer overflow

            if((mid == 0 || nums[mid] <= nums[mid-1]) && (mid == n-1 || nums[mid] <= nums[mid+1])){
                return nums[mid];
            }else if(nums[low] <= nums[mid]){
                low = mid + 1;
            }else{
                high = mid -1;
            }

        }

        return -1;
    }

    public static void main(String[] args) {
        Solution3 s = new Solution3();
        int arr[] = {4,5,6,7,0,1,2};
        System.out.println(s.findMin(arr));
    }
}